website/src/comment-panel.tsx

```diff
@@ -0,0 +1,108 @@
+import { useState } from "react"
+import * as Dailp from "src/graphql/dailp"
+import * as css from "./comment-panel.css"
+import { Button } from "./components"
+import { TranslatedParagraph } from "./segment"
+
+export const CommentPanel = (p: {
+  segment: Dailp.FormFieldsFragment | TranslatedParagraph
+  setIsCommenting: React.Dispatch<SetStateAction<boolean>>
+}) => {
+  const [newCommentText, setNewCommentText] = useState<string>("")
+  const [newCommentType, setNewCommentType] = useState<string>("STORY")
+
+  // Prob messing up smth here
+  const [postCommentResult, postComment] = Dailp.usePostCommentMutation()
+
+  const commentTypeNames: Record<string, Dailp.CommentType> = {
+    // ... TS will then make sure you have an entry for everything on the "CommentTag" type that you import from the codegen
+    Story: Dailp.CommentType.Story,
+    Suggestion: Dailp.CommentType.Suggestion,
+    Question: Dailp.CommentType.Question,
+  }
+
+  /** Call the backend GraphQL mutation. */
+  const runUpdate = async (variables: { input: Dailp.PostCommentInput }) => {
+    return await postComment(variables)
+  }
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    setNewCommentText(event.target.value)
+  }
+
+  const handleSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setNewCommentType(event.target.value)
+  }
+
+  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
+    e.preventDefault()
+
+    if (newCommentText && newCommentType) {
+      runUpdate({
+        input: {
+          parentId: p.segment.id,
+          parentType:
+            p.segment.__typename === "DocumentParagraph"
+              ? Dailp.CommentParentType.Paragraph
+              : Dailp.CommentParentType.Word,
+          textContent: newCommentText,
+          commentType: newCommentType as Dailp.CommentType,
+        },
+      })
+      alert("Your comment has been posted!")
+      console.log("Submitted!")
+      p.setIsCommenting(false)
+    } else {
+      alert("Please add a comment before submitting.")
+    }
+  }
+
+  return (
+    <div>
+      <header className={css.wordPanelHeader}>
+        <h2 className={css.editCherHeader}>
+          {p.segment.__typename === "DocumentParagraph"
+            ? "Paragraph " + p.segment?.index
+            : p.segment.source}
+        </h2>
+      </header>
+      <form>
+        <textarea
+          placeholder="Add a comment..."
+          value={newCommentText}
+          onChange={handleInputChange}
+          className={css.inputStyling}
+        />
+        <div>
+          <label htmlFor="dropdown" className={css.spacing}>
+            Tag:
+          </label>
+          <select
+            id="dropdown"
+            value={newCommentType}
+            onChange={handleSelectChange}
+            className={css.spacing}
+          >
+            {Object.entries<Dailp.CommentType>(commentTypeNames).map(
```

Comment:
@CharlieMcVicker Re: is this what you were envisioning for storing and displaying the commentType names? I will say while implementing this, it felt substantially more complicated and less intuitive for any improvement in functionality that I could understand-- since I think we would still have to go manually update the Record if we wanted to at some point add more commentTypes. I also could have just misunderstood what you were suggesting, though

Reply:
This is what I was thinking (expect that `[label, option]` will flip when you flip the record). I think that even though we **still have to go manually update the Record**, that's actually necessary. If somebody adds a new tag, they shouldn't have to assume that what they put in that enum is going to be directly displayed to a user. That's not what it means to be in an enum, and so we can't assume that someone editing the enum knows anything about how stuff gets displayed. And we actually probably don't want them to have to think about that.

Now, if someone makes a change to the enum (ie. they add a tag), they will get an error that says "This object called `commentTypeNames` promised to have a user-facing name for every comment type, but it's missing a field for _your new type_." I think that should pretty quickly point them in the direction of "Oh I should add a user facing name." IMO that's sort of the best we can hope for in this case.

If we did something clever like map the user-facing names from the enum values we could have something happen like:
1. A developer adds a new tag with some weird developer facing name `COMMENT_SUGGESTION_REPLY_V2` or something. Idk people name stuff weird
2. The code compiles fine and so it gets merged
3. Somebody encounters this weird developer speak and issues a bug
4. We notice the names are mapped from the enum and so we'd have to change the enum
5. Values of `COMMENT_SUGGESTION_REPLY_V2` are already in the database so we have also have to run a database migration to change `COMMENT_SUGGESTION_REPLY_V2` to something more human-readable
6. Three weeks later, we decide we don't like the name we picked and we want to change it again, and we have to update the schema and run a migration, again.

Reply:
Thanks so much for this explanation! I understand now and this makes sense to me